crates/flowctl/src/raw/oauth.rs

```diff
@@ -0,0 +1,273 @@
+
+    #[derive(serde::Deserialize)]
+    struct AuthorizeResponse {
+        url: String,
+        state: String,
+        code_verifier: String,
+    }
+
+    let authorize_response = reqwest::Client::new()
+        .post(oauth_endpoint.clone())
+        .bearer_auth(api.access_token.to_owned())
+        .header("apikey", api.public_token.to_owned())
+        .json(&serde_json::json!({
+            "operation": "auth-url",
+            "connector_config": {
+                "oauth2_client_id": client_id,
+                "oauth2_spec": oauth_spec
+            },
+            "redirect_uri": format!("http://localhost:{port}").as_str(),
+            "config": endpoint_config
+        }))
+        .send()
+        .await
+        .context("Fetching auth-url")?
+        .json::<AuthorizeResponse>()
+        .await
+        .context("Parsing auth-url response")?;
+
+    tracing::info!(url = authorize_response.url, "Opening authorize URL");
+
+    open::that(authorize_response.url)?;
+
+    let redirect_params = get_single_request_query(([127, 0, 0, 1], port)).await?;
+
+    tracing::info!(
+        redirect_params = ?redirect_params,
+        "Got auth code response parameters"
+    );
+
+    let mut code_request_body = serde_json::json!({
+        "operation": "access-token",
+        "connector_config": {
+            "oauth2_client_id": client_id,
+            "oauth2_client_secret": client_secret,
+            "oauth2_injected_values": injected_values.as_ref().unwrap_or(&serde_json::json!(null)),
+            "oauth2_spec": oauth_spec
+        },
+        "state": authorize_response.state,
+        "code_verifier": authorize_response.code_verifier,
+        "config": endpoint_config
+    });
+
+    let code_request_map = code_request_body.as_object_mut().unwrap();
+
+    for (k, v) in redirect_params.into_iter() {
+        code_request_map.insert(k, serde_json::Value::String(v));
+    }
+
+    tracing::info!("Exchanging auth code for access token");
+
+    let code_response = reqwest::Client::new()
+        .post(oauth_endpoint)
+        .bearer_auth(api.access_token.to_owned())
+        .header("apikey", api.public_token.to_owned())
+        .json(&code_request_body)
+        .send()
+        .await?
+        .json::<serde_json::Value>()
+        .await?;
+
+    tracing::info!(
+        "🎉 Got access token response: \n{}",
+        serde_json::to_string_pretty(&code_response)?
+    );
+
+    drop(guard);
+
+    Ok(())
+}
+
+/// Spawns an HTTP server that listens for a single request
+/// then shuts down and returns that requet's query params.
+pub async fn get_single_request_query<S>(listen_addr: S) -> anyhow::Result<HashMap<String, String>>
+where
+    std::net::SocketAddr: From<S>,
+    S: Debug,
+{
+    let (tx, rx) = tokio::sync::oneshot::channel::<HashMap<String, String>>();
+    // All this because a oneshot Sender's `.send()` method consumes `self`
+    let wrapped = Arc::new(Mutex::new(Some(tx)));
+    let service = warp::get()
```

Comment:
Another optional/informational nit pick, but I've elsewhere found that `axum` is easier to use and has a nicer ecosystem. Kind of a non-issue since we no longer have other rust-based web servers in this repo.

Reply:
Yeah, I tried to find our best-practice and realized we didn't really have one. Unless you think it's a big deal I'm inclined to leave it